### PR TITLE
dsaControl module: Reject bids without meta.dsa when required

### PR DESCRIPTION
**Pull request by dgirardi**
## Type of change

- [x] Feature

## Description of change

Update ortbConverter to to carry ORTB response `seatbid[].bid[].ext.dsa` into bid responses' `meta.dsa`; also, add a new `dsaControl` module that:

 - rejects bids that do not provide `meta.dsa` if the auction had `ortb2.regs.ext.dsa.required` set to either 2 ("Required, bid responses without DSA object will not be accepted") or 3 ("Required, bid responses without DSA object will not be accepted, Publisher is an Online Platform")
 - logs a warning for bids that do not provide `meta.dsa` if required was set to 1 ("Supported, bid responses with or without DSA object will be accepted")

## Other information

Closes https://github.com/prebid/Prebid.js/issues/10957
ORTB DSA reference: https://github.com/InteractiveAdvertisingBureau/openrtb/blob/main/extensions/community_extensions/dsa_transparency.md

### CODE REVIEW

**Comment by robertrmartinez** [comment]
So we will not be adding DSA by default to the standard ORTB Converter?

Only when the new module is included?

So - DSA will not be attached unless pub includes dsaControl module in their bundle.

Even if a bidder responds with it (unless they add DSA overrides to their module / converter)

I thought that DSA would always be included in standard ORTB Converter?

File: modules/dsaControl.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  toggleHooks(cfg.consentManagement?.dsa?.validateBids ?? true);
+});
+
+export function setMetaDsa(bidResponse, bid) {
+  if (bid.ext?.dsa) {
+    bidResponse.meta.dsa = bidResponse.meta.dsa ?? bid.ext.dsa;
+  }
+}
+
+registerOrtbProcessor({
```
Comment (patmmccann) [inline modules/dsaControl.js:60]: should this just be in core? It seems if not we'll be in a world where dsa info being in meta or not will be conditional on if you install this module for only a subset of your partners

**Comment by dgirardi** [comment]
I remember discussing this multiple times and agreeing that we want a separate module because non-eu people do not care about dsa.

**Comment by robertrmartinez** [comment]
I have been a flake in the meetings lately so was not there.

So probably always include DSA module when you are going to include GDPR modules.

**Comment by patmmccann** [comment]
> So probably always include DSA module when you are going to include GDPR modules.

This seems wild, many many publishers do not have DSA apply that have european users

**Comment by robertrmartinez** [comment]
I did not mean prebid will auto include, I mean more as a decision pubs / providers make to include it when they do GDPR.

So what is our conclusion here?

All DSA logic in its own optional module?

**Comment by dgirardi** [comment]
Currently, with this PR, ortbConverter sets meta.dsa universally; but rejection happens only if you include the dsaControl module.

**Review by patmmccann — APPROVED** [review]

**Comment by patmmccann** [comment]
required is changing names https://github.com/InteractiveAdvertisingBureau/openrtb/pull/158/files

**Comment by patmmccann** [comment]
Also possible to toss: if pubrender = 0, toss the response if the advertiser cannot render

**Comment by patmmccann** [comment]
linking https://github.com/prebid/prebid-server/issues/3367 as parallel effort in the other repo

**Comment by patmmccann** [comment]
Additional requirement:  Filter responses based on consent to purposes from this table 

image

**Comment by dgirardi** [comment]
> Also possible to toss: if pubrender = 0, toss the response if the advertiser cannot render

How about when pubrender = 2 ("Publisher will render"), but the response has adrender = 1 ("buyer/advertiser will render") ?

**Comment by dgirardi** [comment]
About consent checks - translated into activity controls as they are now, it would mean that when a response has `dsaparams` containing:

- `1` ("Profiling") - we check both `transmitUfpd` and `transmitEids`
- `2` ("Basic ads") - we check `fetchBids`
- `3` ("Precise geolocation") - we check `transmitPreciseGeo`

However, those activities are not meant to be used when _receiving_ bids, as their names indicate. I am worried that they do not fit the use case; if I deny a particular bidder permission to use my user first party data, I am not necessarily denying them the ability to do user profiling through other means. 

I can see a scenario where TCF purpose 4 is *not* denied, but `transmitEids`/`transmitUfpd` are;  the SSP still does some profiling through for example referrers and geography, and indicates so in their dsa response. We would then mistakenly toss out the bid.

Do we need new activities?

**Comment by bretg** [comment]
What is JS doing to "toss" the bid responses? What will an analytics adapter see? On the Prebid Server side, we have this shiny new "seat nonbid" system where we can provide analytics adapters a reason why the bid was rejected instead of just "error" or "no bid".

**Comment by dgirardi** [comment]
An analytics adapter would see a `bidRejected` event with one of these two messages:

https://github.com/prebid/Prebid.js/blob/b296b7226441c665470d1def2d356893fe71b7da/src/constants.json#L134-L135

Something similar would happen if we check for TCF purpose consent - this would be the first instance of that causing a bid to be rejected.

**Review by ChrisHuie — APPROVED** [review]

**Pull request merged**